firedrake/slate/slate.py

```diff
@@ -0,0 +1,616 @@
+"""SLATE is a symbolic language defining a framework for
+performing linear algebra operations on finite element
+tensors. It is similar in principle to most linear algebra
+libraries in notation.
+
+The design of SLATE was heavily influenced by UFL, and
+utilizes much of UFL's functionality for FEM-specific form
+manipulation.
+
+Unlike UFL, however, once forms are assembled into SLATE
+`Tensor` objects, one can utilize the operations defined
+in SLATE to express complicated linear algebra operations.
+(Such as the Schur-complement reduction of a block-matrix
+system)
+
+All SLATE expressions are handled by a specialized linear
+algebra compiler, which interprets SLATE expressions and
+produces C++ kernel functions to be executed within the
+Firedrake architecture.
+"""
+from __future__ import absolute_import, print_function, division
+
+from firedrake.utils import cached_property
+
+from ufl.form import Form
+from ufl.algorithms.map_integrands import map_integrand_dags
+from ufl.algorithms.multifunction import MultiFunction
+from ufl.domain import join_domains, sort_domains
+
+
+__all__ = ['TensorBase', 'Tensor', 'Inverse', 'Transpose',
+           'UnaryOp', 'Negative', 'BinaryOp', 'TensorAdd',
+           'TensorSub', 'TensorMul']
+
+
+class TensorBase(object):
+    """An abstract SLATE node class.
+
+    .. warning::
+
+       Do not instantiate this class on its own. This is an abstract
+       node class; is not meant to be worked with directly. Only use
+       the appropriate subclasses.
+    """
+
+    id = 0
+
+    def __init__(self):
+        """Constructor for the TensorBase abstract class."""
+        self.id = TensorBase.id
+        TensorBase.id += 1
+        self._integral_domains = None
+        self._subdomain_data = None
+        self._hash = None
+
+    @cached_property
+    def shape(self):
+        """Computes the shape information of the local tensor."""
+        shape = []
+        for i, arg in enumerate(self.arguments()):
+            V = arg.function_space()
+            shapelist = []
+            for fs in V:
+                shapelist.append(fs.fiat_element.space_dimension() * fs.dim)
+            shape.append(sum(shapelist))
```

Comment:
You could probably reduce some code duplication if you calculate `shape` from `shapes`.